vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,907 @@
+    }
+    else
+    {
+      labeled_example = false;
+    }
+  }
+
+  bool holdout_example = labeled_example;
+  for (size_t i = 0; i < ec_seq.size(); i++) holdout_example &= ec_seq[i]->test_only;
+
+  // TODO what does weight mean here?
+  all.sd->update(holdout_example, labeled_example, loss, ec_seq[SHARED_EX_INDEX]->weight, num_features);
+
+  for (auto sink : all.final_prediction_sink)
+    print_decision_scores(sink, ec_seq[SHARED_EX_INDEX]->pred.decision_scores);
+
+  CCB::print_update(all, slots, preds, num_features);
+}
+
+void finish_multiline_example(vw& all, ccb& data, multi_ex& ec_seq)
+{
+  if (ec_seq.size() > 0)
+  {
+    output_example(all, data, ec_seq);
+    CB_ADF::global_print_newline(all);
+  }
+
+  for (auto a_s : ec_seq[0]->pred.decision_scores)
+  {
+    a_s.delete_v();
+  }
+  ec_seq[0]->pred.decision_scores.delete_v();
+
+  // Delete all of the labels originally allocated by the parser.
+  for (auto ex : ec_seq)
+  {
+    all.p->lp.delete_label(&ex->l);
+  }
+
+  VW::clear_seq_and_finish_examples(all, ec_seq);
+}
+
+void finish(ccb& data)
+{
+  data.actions.~vector<example*>();
+  data.slots.~vector<example*>();
+  data.origin_index.~map<uint32_t,uint32_t>();
+  data.exclude_list.~vector<bool>();
+  // data.include_list.~unordered_set<uint32_t>();
+  data.cb_label.~cb_class();
+  data.default_cb_label.~cb_class();
+  data.generated_interactions.~vector<std::string>();
+  data.stored_labels.~vector<CCB::label>();
+}
+
+base_learner* ccb_explore_adf_setup(options_i& options, vw& all)
+{
+  auto data = scoped_calloc_or_throw<ccb>();
+  bool ccb_explore_adf_option = false;
+  option_group_definition new_options("Conditional Contextual Bandit Exploration with Action Dependent Features");
+  new_options.add(make_option("ccb_explore_adf", ccb_explore_adf_option)
+                      .keep()
+                      .help("Do Conditional Contextual Bandit learning with multiline action dependent features."));
+  options.add_and_parse(new_options);
+
+  if (!ccb_explore_adf_option)
+    return nullptr;
+
+  if (!options.was_supplied("cb_explore_adf"))
+  {
+    options.insert("cb_explore_adf", "");
+    options.add_and_parse(new_options);
+  }
+
+  if (!options.was_supplied("cb_sample"))
+  {
+    options.insert("cb_sample", "");
+    options.add_and_parse(new_options);
+  }
+
+  auto base = as_multiline(setup_base(options, all));
+  all.p->lp = CCB::ccb_label_parser;
+  all.label_type = label_type::ccb;
+
+  // Extract from lower level reductions
+  data->default_cb_label = {FLT_MAX, 0, -1.f, 0.f};
+  data->shared = nullptr;
+  data->original_interactions = &all.interactions;
+
+  learner<ccb, multi_ex>& l =
+      init_learner(data, base, learn_or_predict<true>, learn_or_predict<false>, 1, prediction_type::decision_probs);
+
+  l.set_finish_example(finish_multiline_example);
+  l.set_finish(CCB::finish);
+  return make_base(l);
+}
+
+size_t read_cached_label(shared_data*, void* v, io_buf& cache)
```

Comment:
Maybe we should separate the ccb label type stuff from the ccb reduction?  That seems conceptually sound and agrees with how other reductions are done.

Reply:
They are separate, just happen to be in the same file. I agree, this file is messy because of it. Will move